spec/function.dd

```diff
@@ -2952,6 +2960,227 @@ $(H3 $(LNAME2 system-functions, System Functions))
+                $(SPEC_RUNNABLE_EXAMPLE_COMPILE
+                ---
+                @safe void memcpy(E)(E[] src, E[] dst)
+                {
+                    import std.math : min;
+                    foreach (i; 0 .. min(src.length, dst.length))
+                    {
+                        dst[i] = src[i];
+                    }
+                }
+                ---
+                )
+                because the rules for safe $(RELATIVE_LINK2 safe-values, safe
+                values) ensure that the lengths of the arrays are correct.
+            )
+        )
+
+$(H3 $(LNAME2 safe-values, Safe Values))
+
+        $(P For $(DDSUBLINK spec/type, basic-data-types, basic data types), all
+        possible bit patterns are safe.)
+
+        $(P A pointer is safe when:)
+        $(OL
+            $(LI it can be dereferenced validly, and)
+            $(LI the value of the pointee is safe.)
+        )
+        $(P Examples:)
+        $(SPEC_RUNNABLE_EXAMPLE_RUN
+        ---
+        int* n = null; /* n is safe because dereferencing null is a well-defined
+            crash. */
+        int* x = cast(int*) 0xDEADBEEF; /* x is (most likely) unsafe because it
+            is not a valid pointer and cannot be dereferenced. */
+
+        import core.stdc.stdlib: malloc, free;
+        int* p1 = cast(int*) malloc(int.sizeof); /* p1 is safe because the
+            pointer is valid and *p1 is safe regardless of its actual value. */
+        free(p1); /* This makes p1 unsafe. */
+        int** p2 = &p1; /* While it can be dereferenced, p2 is unsafe because p1
+            is unsafe. */
+        p1 = null; /* This makes p1 and p2 safe. */
+        ---
+        )
+
+        $(P A dynamic array is safe when:)
+        $(OL
+            $(LI its pointer is safe, and)
+            $(LI its length is in-bounds with the corresponding memory object,
+            and)
+            $(LI all its elements are safe.)
+        )
+
+        $(P Examples:)
+        $(SPEC_RUNNABLE_EXAMPLE_COMPILE
+        ---
+        int[] f() @system
+        {
+            int[3] a;
+            int[] d1 = a[0 .. 2]; /* d1 is safe. */
+            int[] d2 = a.ptr[0 .. 3]; /* d2 is unsafe because it goes beyond a's
+                bounds. */
+            int*[] d3 = [cast(int*) 0xDEADBEEF]; /* d3 is unsafe because the
+                element is unsafe. */
+            return d1; /* Up to here, d1 was safe, but its pointer becomes
+                invalid when the function returns, so the returned dynamic array
+                is unsafe. */
+        }
+        ---
+        )
+
+        $(P A static array is safe when all its elements are safe. Regardless
+        of the element type, a static array with length zero is always safe.)
+
+        $(P An associative array is safe when all its keys and elements are
+        safe.)
+
+        $(P A struct/union instance is safe when:)
+        $(OL
+            $(LI the values of its accessible fields are safe, and)
```

Comment:
> What counts as a non-accessible field?

It's meant in terms of `public`, `private`, etc. For example, a `private` field is not accessible, unless your code is in the same module. Maybe making "accessible" a link to <https://dlang.org/spec/attribute.html#visibility_attributes> could make this clearer?

This provides some support for the current practice of using `private` to hide unsafe implementations. With this rule, at least using those types is kosher. Of course, the implementations are still breaking the rules. Which is where `@system` variables come in. And when those land, it might make sense to revisit this.